src/pretrain_utils.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2021 Google LLC
+# 
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# 
+#     https://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import attr
+from typing import Text, Mapping, List
+
+import tensorflow as tf
+import tensorflow_text as tf_text
+
+from google_research.etcmodel.models.modeling import EtcConfig 
+from input_utils import PretrainInputConfig, get_pretrain_example_decode_fn
+
+
+def input_fn_builder(tokenizer: tf_text.BertTokenizer,
+                     input_filenames: List,
+                     input_config: PretrainInputConfig,
+                     model_config: EtcConfig,
+                     is_training: bool,
+                     num_cpu_threads: int = 4):
+  """Creates an `input_fn` closure to be passed to TPUEstimator."""
+
+  vocab = tokenizer._wordpiece_tokenizer._get_vocab_and_ids()[0]
+  vocab = vocab.numpy().tolist()
+  # [PATCH] is b'[unused0]'
+  unselectable_tokens = [b'[CLS]', b'[SEP]', b'[unused0]']
+  unselectable_tokens += [f'[unused{i}]'.encode() 
```

Comment:
What are these additional unselectable_tokens used for?

Reply:
Unused tokens are used as special tokens that indicate the types of subsequences. For example, [unused0] is for [PATCH] which will be dded prior to the sequence of patch tokens.

Reply:
Yes, but I meant Line 37.

Reply:
Unused tokens from [unused1] to [unusedN] will be used as special tokens for text. N will be determined by the number of text fields we use; in other words, N will be len(input_config.text_keys).

Reply:
I see. Interesting. Thanks Roy!